internal/cli/command/cluster/update.go

```diff
@@ -0,0 +1,158 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"emperror.dev/errors"
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/Masterminds/semver"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/banzaicloud/banzai-cli/.gen/pipeline"
+	"github.com/banzaicloud/banzai-cli/internal/cli"
+	clustercontext "github.com/banzaicloud/banzai-cli/internal/cli/command/cluster/context"
+	"github.com/banzaicloud/banzai-cli/internal/cli/format"
+	"github.com/banzaicloud/banzai-cli/internal/cli/utils"
+)
+
+type updateOptions struct {
+	file     string
+	wait     bool
+	interval int
+	clustercontext.Context
+}
+
+// NewUpdateCommand creates a new cobra.Command for `banzai cluster update`.
+func NewUpdateCommand(banzaiCli cli.Cli) *cobra.Command {
+	options := updateOptions{}
+
+	cmd := &cobra.Command{
+		Use:          "update",
+		Aliases:      []string{"u"},
+		Short:        "Update a cluster",
+		Long:         "Update cluster based on json stdin or interactive session",
+		SilenceUsage: true,
+		Args:         cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runUpdate(banzaiCli, options)
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVarP(&options.file, "file", "f", "", "Cluster update descriptor file")
+	flags.BoolVarP(&options.wait, "wait", "w", false, "Wait for cluster update")
+	flags.IntVarP(&options.interval, "interval", "i", 10, "Interval in seconds for polling cluster status")
```

Comment:
**Note**: I might have suggested a `cluster-name` positional argument here similarly to how `banzai cluster get` and `banzai cluster nodepool update` is implemented.

Also I'm not sure whether we expect future updates to contain more parameters but as long as only a version is expected I feel like the descriptor file seems to be an overkill, I would be glad to specify the version with a flag or interactively - of which the latter is available.

Reply:
There are a bunch of other parameters that we will have to support eventually

Reply:
```go
	options.Context = clustercontext.NewClusterContext(cmd, banzaiCli, "update")
```
"installs" these flags to the subcommand automatically.

Reply:
Regarding the version flag: that makes sense, it's good this way then.

Regarding the positional parameter: yes, I checked that, what I meant by my first paragraph was that the current implementation seems to support only explicit `--cluster-name <name>` or `--cluster-id <id>` calls, but the following would be the most intuitive for me:
`banzai cluster update my-little-cluster --file ~/banzai-cluster-update.yaml` 
instead of
`banzai cluster update --cluster-name my-little-cluster --file ~/banzai-cluster-update.yaml`
similarly to other commands of ours such as `cluster get` and `cluster nodepool update`.
It's a small thing and can be added later, it just makes the CLI more consistent IMO.

```shell
pregnor_41568:Pregnor-BanzaiCloud-MBP152019@194.39.44.200/Users/pregnor/development/src/github.com/banzaicloud/banzai-cli (master u=)
> build/banzai cluster update --help

# Begin: 2020-09-01T16:36:08.121+02:00

Update cluster based on json stdin or interactive session

Usage:
  banzai cluster update [flags]

Aliases:
  update, u

Flags:
      --cluster int32         ID of cluster to update
      --cluster-name string   Name of cluster to update
  -f, --file string           Cluster update descriptor file
  -h, --help                  help for update
  -i, --interval int          Interval in seconds for polling cluster status (default 10)
  -w, --wait                  Wait for cluster update

Global Flags:
      --color                use colors on non-tty outputs
      --config string        config file (default is $BANZAICONFIG or $HOME/.banzai/config.yaml)
      --interactive          ask questions interactively even if stdin or stdout is non-tty
      --no-color             never display color output
      --no-interactive       never ask questions interactively
      --organization int32   organization id
  -o, --output string        output format (default|yaml|json) (default "default")
      --verbose              more verbose output

# End: 2020-09-01T16:36:08.223+02:00
# Duration: 102ms
# Exit code: 0
```

```shell
pregnor_41568:Pregnor-BanzaiCloud-MBP152019@194.39.44.200/Users/pregnor/development/src/github.com/banzaicloud/banzai-cli (master u=)
> build/banzai cluster get --help

# Begin: 2020-09-01T16:38:51.385+02:00

Get cluster details

Usage:
  banzai cluster get [--cluster=ID | [--cluster-name=]NAME] [flags]

Aliases:
  get, g, show

Flags:
      --cluster int32         ID of cluster to get
      --cluster-name string   Name of cluster to get
  -h, --help                  help for get

Global Flags:
      --color                use colors on non-tty outputs
      --config string        config file (default is $BANZAICONFIG or $HOME/.banzai/config.yaml)
      --interactive          ask questions interactively even if stdin or stdout is non-tty
      --no-color             never display color output
      --no-interactive       never ask questions interactively
      --organization int32   organization id
  -o, --output string        output format (default|yaml|json) (default "default")
      --verbose              more verbose output

# End: 2020-09-01T16:38:51.490+02:00
# Duration: 105ms
# Exit code: 0
```

```shell
pregnor_41568:Pregnor-BanzaiCloud-MBP152019@194.39.44.200/Users/pregnor/development/src/github.com/banzaicloud/banzai-cli (master u=)
> build/banzai cluster nodepool update --help

# Begin: 2020-09-01T16:39:16.028+02:00

Update a node pool (and related subcommands)

Usage:
  banzai cluster nodepool update [NAME] [flags]
  banzai cluster nodepool update [command]

Aliases:
  update, u, upgrade

Available Commands:
  cancel      Cancel a node pool update
  tail        Tail a node pool update

Flags:
      --cluster int32         ID of cluster to update
      --cluster-name string   Name of cluster to update
  -f, --file string           Node pool descriptor file
  -h, --help                  help for update

Global Flags:
      --color                use colors on non-tty outputs
      --config string        config file (default is $BANZAICONFIG or $HOME/.banzai/config.yaml)
      --interactive          ask questions interactively even if stdin or stdout is non-tty
      --no-color             never display color output
      --no-interactive       never ask questions interactively
      --organization int32   organization id
  -o, --output string        output format (default|yaml|json) (default "default")
      --verbose              more verbose output

Use "banzai cluster nodepool update [command] --help" for more information about a command.

# End: 2020-09-01T16:39:16.073+02:00
# Duration: 45ms
# Exit code: 0
```

Reply:
speaking about consistency, we should also consider that we couldn't update a cluster named 'tail' or 'cancel' this way :)